src/metric_engine/src/index/cache.rs

```diff
@@ -0,0 +1,1293 @@
+type ConcurrentTagKVMap =
+    RwLock<HashMap<TagName, HashMap<TagValue, HashMap<MetricId, HashSet<SeriesId>>>>>;
+
+struct MetricsCache {
+    cache: DashMap<SegmentDuration, ConcurrentMetricMap>,
+    pub storage: TimeMergeStorageRef,
+    sender: Sender<Task>,
+}
+struct SeriesCache {
+    cache: DashMap<SegmentDuration, ConcurrentSeriesMap>,
+    pub storage: TimeMergeStorageRef,
+    sender: Sender<Task>,
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+struct SegmentSeries {
+    segment: SegmentDuration,
+    series_id: SeriesId,
+}
+
+struct TagIndexCache {
+    cache: DashMap<SegmentDuration, ConcurrentTagKVMap>,
+    series_records: RwLock<HashSet<SegmentSeries>>,
+    storage: TimeMergeStorageRef,
+    sender: Sender<Task>,
+}
+
+impl MetricsCache {
+    fn new(storage: TimeMergeStorageRef, sender: Sender<Task>) -> Self {
+        Self {
+            cache: DashMap::new(),
+            storage,
+            sender,
+        }
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn parse_record_batch(
+        batch: &RecordBatch,
+        index: usize,
+    ) -> Result<(&[u8], &[u8], u8, u64, u64, u64)> {
+        let metric_name = batch
+            .column_by_name(COLUMN_METRIC_NAME)
+            .context("get column failed")?
+            .as_any()
+            .downcast_ref::<BinaryArray>()
+            .context("parse column failed")?
+            .value(index);
+
+        let field_name = batch
+            .column_by_name(COLUMN_FIELD_NAME)
+            .context("get column failed")?
+            .as_any()
+            .downcast_ref::<BinaryArray>()
+            .context("parse column failed")?
+            .value(index);
+
+        let field_type = batch
+            .column_by_name(COLUMN_FIELD_TYPE)
+            .context("get column failed")?
+            .as_any()
+            .downcast_ref::<UInt8Array>()
+            .context("parse column failed")?
+            .value(index);
+
+        let filed_id = batch
+            .column_by_name(COLUMN_FIELD_ID)
+            .context("get column failed")?
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .context("parse column failed")?
+            .value(index);
+
+        let metric_id = batch
+            .column_by_name(COLUMN_METRIC_ID)
+            .context("get column failed")?
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .context("parse column failed")?
+            .value(index);
+
+        let duration = batch
+            .column_by_name(COLUMN_DURATION)
+            .context("get column failed")?
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .context("parse column failed")?
+            .value(index);
+
+        Ok((
+            metric_name,
+            field_name,
+            field_type,
+            filed_id,
+            metric_id,
+            duration,
+        ))
+    }
+
+    async fn load_from_storage(&mut self) -> Result<()> {
```

Comment:
The cache module shouldn't known how the persistence layer is implemented, we can move this method to `IndexManager`.